lib/utilities/will-interrupt-process.js

```diff
@@ -1,17 +1,68 @@
 'use strict';
 
-// (un)Register process interruption signals(SIGINT, SIGTERM, process.kill()) handlers.
+// Allows to setup process interruption handlers.
+// The process can be interrupted when ges SIGINT, SIGTERM signal,
+// something called process.exit(exitCode) or CTRL+C was pressed.
 //
-// When the process interrupted("CTR+C") all the registered handlers are being invoked via `capture-exit`.
-// The process finishes after the last interruption handler is settled.
+// Node.js doesn't allow to perform async tasks when the process is exiting.
+// Also there are some work arounds for exit in the node.js ecosystem.
+//
+// In order to supply reliable process exit phase, `will-interrupt-process`
+// is tightly integrated with `capture-exit` which allows us to perform async cleanup
+// on `process.exit()` and control the final exit code.
+
+const captureExit = require('capture-exit');
+const EventEmitter = require('events');
 
-let exit = require('capture-exit');
-exit.captureExit();
+const handlers = [];
 
-let windowsCtrlCTrap;
-let handlers = [];
+let _process,
+    windowsCtrlCTrap;
 
 module.exports = {
+  capture(outerProcess) {
+    if (_process) {
+      throw new Error('process already captured');
+    }
+
+    if (outerProcess instanceof EventEmitter === false) {
+      throw new Error('attempt to capture bad process instance');
+    }
+
+    _process = outerProcess;
+
+    // ember-cli and user apps have many dependencies, many of which require
+    // process.addListener('exit', ....) for cleanup, by default this limit for
+    // such listeners is 10, recently users have been increasing this and not to
+    // their fault, rather they are including large and more diverse sets of
+    // node_modules.
+    //
+    // https://github.com/babel/ember-cli-babel/issues/76
+    _process.setMaxListeners(1000);
+
+    // work around misbehaving libraries, so we can correctly cleanup before actually exiting.
+    captureExit.captureExit();
+  },
+
+  /**
+   * Drops all the interruption handlers and disables an ability to add new one
+   *
+   * Note: We don't call `captureExit.releaseExit() here.
+   * In some rare scenarios it can lead to the hard to debug issues.
```

Comment:
here I'm referring to the point 9) from [this comment](https://github.com/ember-cli/ember-cli/issues/6779#issuecomment-280940358). Not sure how to better document this